quantum/orgb.c

```diff
@@ -0,0 +1,349 @@
+            raw_hid_buffer[1] = ORGB_FAILURE;
+            raw_hid_buffer[2] = ORGB_EOM;
+            continue;
+        }
+
+        g_orgb_direct_mode_colors[first_led + i].r = r;
+        g_orgb_direct_mode_colors[first_led + i].g = g;
+        g_orgb_direct_mode_colors[first_led + i].b = b;
+        i++;
+    }
+
+    raw_hid_buffer[1] = ORGB_SUCCESS;
+    raw_hid_buffer[2] = ORGB_EOM;
+}
+
+void orgb_set_mode(uint8_t *data) {
+    const bool    should_save_to_eeprom = data[1] == ORGB_EEPROM_SAVE;
+    const uint8_t mode                  = data[2];
+    if (mode >= RGB_MATRIX_EFFECT_MAX) {
+        raw_hid_buffer[1] = ORGB_FAILURE;
+        raw_hid_buffer[2] = ORGB_EOM;
+        return;
+    }
+
+    if (should_save_to_eeprom)
+        rgb_matrix_mode(mode);
+    else
+        rgb_matrix_mode_noeeprom(mode);
+
+    raw_hid_buffer[1] = ORGB_SUCCESS;
+    raw_hid_buffer[2] = ORGB_EOM;
+}
+void orgb_set_mode_and_speed(uint8_t *data) {
+    const bool    should_save_to_eeprom = data[1] == ORGB_EEPROM_SAVE;
+    const uint8_t mode                  = data[2];
+    const uint8_t speed                 = data[3];
+
+    if (mode > RGB_MATRIX_EFFECT_MAX || speed > 255) {
+        raw_hid_buffer[1] = ORGB_FAILURE;
+        raw_hid_buffer[2] = ORGB_EOM;
+        return;
+    }
+
+    if (should_save_to_eeprom) {
+        rgb_matrix_mode(mode);
+        rgb_matrix_set_speed(speed);
+    } else {
+        rgb_matrix_mode_noeeprom(mode);
+        rgb_matrix_set_speed_noeeprom(speed);
+    }
+
+    raw_hid_buffer[1] = ORGB_SUCCESS;
+    raw_hid_buffer[2] = ORGB_EOM;
+}
+void orgb_set_color_mode_and_speed(uint8_t *data) {
+    const bool    should_save_to_eeprom = data[1] == ORGB_EEPROM_SAVE;
+    const uint8_t h                     = data[2];
+    const uint8_t s                     = data[3];
+    const uint8_t v                     = data[4];
+    const uint8_t mode                  = data[5];
+    const uint8_t speed                 = data[6];
+
+    if (h > 255 || s > 255 || v > 255 || mode > RGB_MATRIX_EFFECT_MAX || speed > 255) {
+        raw_hid_buffer[1] = ORGB_FAILURE;
+        raw_hid_buffer[2] = ORGB_EOM;
+        return;
+    }
+
+    if (should_save_to_eeprom) {
+        rgb_matrix_sethsv(h, s, v);
+        rgb_matrix_mode(mode);
+        rgb_matrix_set_speed(speed);
+    } else {
+        rgb_matrix_sethsv_noeeprom(h, s, v);
+        rgb_matrix_mode_noeeprom(mode);
+        rgb_matrix_set_speed_noeeprom(speed);
+    }
+
+    raw_hid_buffer[1] = ORGB_SUCCESS;
+    raw_hid_buffer[2] = ORGB_EOM;
+}
+
+// Send info to OpenRGB
+void orgb_get_qmk_version(void) {
+    const char *qmk_version = QMK_VERSION;
+    uint8_t     i           = 0;
+    while (qmk_version[i] != 0 && i + 2 < RAW_EPSIZE) {
+        raw_hid_buffer[i + 1] = qmk_version[i];
+        i++;
+    }
+    raw_hid_buffer[i + 1] = ORGB_EOM;
+}
+void orgb_get_device_name(void) {
+    char *name = malloc(strlen(MANUFACTURER) + strlen(" ") + strlen(PRODUCT) + 1);
```

Comment:
Can we use the preprocessor for this, to avoid allocations?
Something like:
```c
static const char name[] = #MANUFACTURER " " #PRODUCT;
```
....or whatever magic is required to stringify them. Some keyboard builds (namely ChibiOS) have allocations disabled completely.

Reply:
`    #define SPACE "  "
    #define name MANUFACTURER SPACE PRODUCT`

That should be better?

Reply:
This would be great, as I had to rework this to get rid of the allocation for the K556 port.

Reply:
It also complained at me that the MANUFACTURER and PRODUCT defines didn't have quotes around them.  Adding quotes causes them to show up as "Redragon" "K556" in lsusb/dmesg.